### PR TITLE
Debug avg planner

### PR DESCRIPTION
**Pull request by Alvinosaur**
Hey Sam, I'm sorry for  leaving all those random txt and deprecated_reference files. I tried to to first make an orphan branch and unstage all files. Then to make sure this branch was comparable to the debug_avg_planner branch, I tried rebasing on top of commit #1, but this just messed things up. Please just ignore those  other files. I've added comments to the relevant files:

main.py
sim_objects.py
environment.py

You can also see that I run experiments by printing out text and redirecting output from stdout to a text file. I also store random things in .npz files.

### CODE REVIEW

File: sim_objects.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        self.fill_prop = np.clip(fill_prop, self.min_fill, self.max_fill)
+        self.bottle_mass = self.mass_from_fill(self.fill_prop)
+        self.center_of_mass = self.com_from_fill(self.fill_prop)
+        self.inertial_shift = self.center_of_mass - self.default_com
+
+    def mass_from_fill(self, fill_prop):
+        return Bottle.PLASTIC_MASS + (
+            fill_prop * self.max_volume * Bottle.VOL_TO_MASS)
+
+    def com_from_fill(self, fill_prop):
```
Comment (SamNPowers) [inline sim_objects.py]: [Minor] "com" -> "center_of_mass" would be clearer for new people reading your code

File: sim_objects.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+            return np.array([0, 0, self.height * 0.4])
+        else:
+            return np.array([0, 0, water_height * 0.4])
+
+    def create_sim_bottle(self, pos=None, ori=None):
+        # if self.bottle_id is not None:
+        #     p.removeBody(self.bottle_id)
+        if ori is None:
+            ori = [0, 0, 0, 1]
+        if pos is not None:
```
Comment (SamNPowers) [inline sim_objects.py]: Unless I'm misreading, you can just do:
```
if pos is None:
    pos = self.start_pos
```

and then you don't need the current if/else (just use pos)

File: sim_objects.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    DEFAULT_HEIGHT = 0.1905  # m
+    DEFAULT_FRIC = 0.1  # plastic-wood dynamic friction
+    DEFAULT_FILL = 0.5
+
+    def __init__(self, start_pos, start_ori,
+                 fill_prop=DEFAULT_FILL,
+                 fric=DEFAULT_FRIC,
+                 radius=DEFAULT_RAD,
+                 height=DEFAULT_HEIGHT):
+        self.start_pos = start_pos
```
Comment (SamNPowers) [inline sim_objects.py:29]: Not sure yet if this applies, but just so you know it's common (and convenient) for private variables to be "\_my\_var" and publicly accessible ones to be "my\_var". I.e. the difference being the starting "\_". (Perhaps you know this and all of these are public.) 

This is convenient for people using your code, to know what to expect - I usually make stuff private by default, unless I have made an explicit decision to make it acceptable for others (other consumers of the class) to use.

File: sim_objects.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+                          (np.linalg.norm(z_axis) * np.linalg.norm(new_z_axis)))
+        # when z-axis rotation is 90deg
+        return abs(angle) > (45 * math.pi / 180)
+
+
+class Arm:
+    def __init__(self, EE_start_pos, start_ori, kukaId):
+        self.EE_start_pos = EE_start_pos
+        self.start_ori = start_ori
+        self.kukaId = kukaId
```
Comment (SamNPowers) [inline sim_objects.py]: [Very Minor] Inconsistent with your snake case everywhere else. (Not gonna make this comment everywhere, but it applies to everything camel case other than class names.)

File: sim_objects.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+        self.base_pos = np.array([0, 0, 0.1])
+        self.min_dist = 0.3
+        self.MAX_HORIZ_REACH = 0.851  # need to set with set_general_max_reach()
```
Comment (SamNPowers) [inline sim_objects.py]: set_general_max_reach() is setting "MAX_REACH" not "MAX_HORIZ_REACH", so not sure if the code is doing something other than what you intended.

By the way, if I *need* something to be set, I'll usually use the @property decorator. I.e. I'll do something along the lines of

```
def __init__(self):
    self._some_var = None

@property
def some_var():
    assert self._some_var is not None, "some_var must be set before it can be used."
    return self._some_var
```

And then set _some_var however you like. (TIL there's a fancy @some_var.setter decorator that can be used, though set_general_max_reach() is fine too)

File: main.py [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    # fill_prop = 0.3
+    # fric = env.max_fric
+    # new_bottle = Bottle(start_pos=bottle_start_pos,
+    #                     start_ori=bottle_start_ori,
+    #                     fill_prop=fill_prop,
+    #                     fric=fric)
+    # env.bottle = new_bottle
+
+    # run planner and visualize result
+    num_sims_per_action = 20
```
Comment (SamNPowers) [inline main.py]: I find it convenient to centralize all my constants, so when I want to change one I don't have to go hunting around for it. I know in academia stuff is more ad-hoc so not really stating this as a problem.

But I do like the pattern you've established of naming the variable before using it, so you're clear on what the number is supposed to mean (vs just using it in-place).

File: sim_objects.py (hunk 3 above)
Comment (SamNPowers) [reply sim_objects.py:29]: By the way this applies to methods too, not just class variables.

**Review by SamNPowers — COMMENTED** [review]
Just spent some time familiarizing myself with the broad strokes of the code, and making some comments as I noticed things I wanted to comment on. I'm not finished, really just starting, but apparently github will lose my "pending" changes if, for instance, my computer runs out of application memory and decides to close my browser (which happens to me not infrequently at the moment), so submitting an unfinished review. 

I'll keep reviewing over the next few days.

**Comment by Alvinosaur** [comment]
> Just spent some time familiarizing myself with the broad strokes of the code, and making some comments as I noticed things I wanted to comment on. I'm not finished, really just starting, but apparently github will lose my "pending" changes if, for instance, my computer runs out of application memory and decides to close my browser (which happens to me not infrequently at the moment), so submitting an unfinished review.
> 
> I'll keep reviewing over the next few days.

Great, thanks Sam! Please take your time, no rush. After the quiz I'll start addressing the comments as well.

File: main.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
+import pybullet as p
```
Comment (SamNPowers) [inline main.py:1]: One thing I'll say here is that code review is partially for your future self (make sure the architecture is good, stuff is tested, etc) but also it's particularly useful such that *other people* can look at and work with your code. So some of my comments will be with that in mind. If you don't care about that use-case, feel free to ignore stuff in that vein. 

Relatedly: It's convenient to have a requirements.txt file so people can pip install all your dependencies at once. Alternatively if you want (others) to be able to pip install your project, you can have a setup.py file. Or a conda yml file to enable conda. Whatever, just some way for people to be able to quickly setup and get started with your project. (It's also possible to have all three from just one list of dependencies, to enable people to use whatever they like without having to maintain several files containing requirements.)

File: main.py [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+                                  ybounds, dist_thresh, eps, da_rad=da_rad,
+                                  dx=dx, dy=dy, dz=dz, visualize=VISUALIZE, sim_mode=NaivePlanner.SINGLE)
+    avg_planner = NaivePlanner(start, goal, env, xbounds,
+                               ybounds, dist_thresh, eps, da_rad=da_rad,
+                               dx=dx, dy=dy, dz=dz, visualize=VISUALIZE, sim_mode=NaivePlanner.AVG,
+                               num_rand_samples=num_sims_per_action)
+
+    # exec_params_set = plan_params_sets[0]
+    single_planner.sim_params_set = plan_params_sets
+    avg_planner.sim_params_set = plan_params_sets[:10]
```
Comment (SamNPowers) [inline main.py]: As-is it seems odd to set these like this, rather than passing into plan() or passing into constructor (not 100% sure yet what these params are used for). Looking further it looks like there's a change_param_set function that should be getting used?

Also it's odd to get these from the environment, pass it into the planner, then have the planner pass them right back into the environment in run_sim? Why the indirection of parameters, rather than just letting environment be the only owners of them? (Or perhaps I'm missing something?)

File: main.py [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+        #     c = i % 4
+        #     plots[r][c].plot(full_arm_traj_policy[:, i], label="Policy traj")
+        #     plots[r][c].plot(full_arm_traj_state[:, i], label="State traj")
+        #     plots[r][c].axvline(x=collision_t, ymin=-5, ymax=5, c='g', label="Collision")
+        #     # plots[r][c].plot(executed_traj[:, i], label="Executed traj")
+
+        #     plots[r][c].set_title("Joint %d" % i)
+
+        # plots[1][2].legend()
+        # plt.show()
```
Comment (SamNPowers) [inline main.py]: I normally say to remove dead code. If you're actively testing stuff out with dead code, that's one thing, but if it sticks around forever it just gets in the way and makes everything harder to read. Plus git keeps everything so it's not like you can't get it back if you want it.

File: main.py [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+
+
+def main():
```
Comment (SamNPowers) [inline main.py]: I would consider breaking this method up into something like:
env = construct_environment()
plan_params = get_plan_params()
planner = create_planner(plan_params)  # Could roll get_plan_params into this, instead
plan = planner.plan()
execute_plan(plan)  # Or perhaps test_plan? It looks like that's more aligned with what the play_results branch above is doing.

You have all the parts here already, it would just be clearer what's happening if it were broken up. Plus then it would be a bit easier to drop in, say, variations of different environments by having different construct_environments() that you are selecting from. Also you could unit test, eg, your execute_plan with some simple test plan in some simple env directly, without first having to use your planner to make it.

File: main.py [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+    start_goal_idx = 10
+    (startb, goalb, start_joints) = start_goals[start_goal_idx]
+    startb = [0.42691895, 0.56569359, 0.04906958]
+    start_joints = [1.00454656, 1.34184254, 2.80407075,
+                    0.80206976, 1.8428638, 2.08643242, 0.1634396]
+    start_state = helpers.bottle_EE_to_state(
+        bpos=startb, arm=arm, joints=start_joints)
+    goal_state = helpers.bottle_EE_to_state(bpos=goalb, arm=arm)
+    planner.start = start_state
+    planner.goal = goal_state
```
Comment (SamNPowers) [inline main.py]: You're passing a start and goal into the planner already. Why set them manually like this? Rather than doing this initialization above and passing the right start/goal in from the beginning?

File: main.py [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+        directly to generate trajectory. Defaults to False.
+
+    Returns:
+        [type]: [description]
+    """
+    cur_joints_policy = np.copy(init_joints)
+    cur_joints_state = np.copy(init_joints)
+    piecewise_trajs_policy = []
+    piecewise_trajs_state = []
+    for i, (dq_vec, num_iters) in enumerate(policy):
```
Comment (SamNPowers) [inline main.py]: "dq" -> words?

File: main.py [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+            start=cur_joints_state, stop=target_joints_state, num=num_iters))
+
+        cur_joints_policy = target_joints_policy
+        cur_joints_state = target_joints_state
+
+    try:
+        if use_policy:
+            full_arm_traj = np.vstack(piecewise_trajs_policy)
+        else:
+            full_arm_traj = np.vstack(piecewise_trajs_state)
```
Comment (SamNPowers) [inline main.py]: full_arm_traj doesn't seem to be getting used?

File: main.py [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+    except Exception as e:
+        print("Couldn't stack piecewise trajectories b/c %s" % e)
```
Comment (SamNPowers) [inline main.py]: You're only really checking one of the stacks, but you then return both from the function? If this stacking is a risk, it seems like you want to do both in the try/except and return the created vals.

File: naive_joint_space_planner.py [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+        # of random env params
+        self.param_probs = [(p.bottle_fill_prob * p.bottle_fric_prob)
+                            for p in self.sim_params_set]
+
+        # safety threshold for prob(bottle fall) to deem invalid transitions
+        self.fall_proportion_thresh = 0.2
+
+        # method of simulating an action
+        self.sim_mode = sim_mode
+        if sim_mode == self.SINGLE:
```
Comment (SamNPowers) [inline naive_joint_space_planner.py]: As-is it doesn't really look like this mode is necessary? Can just use AVG with num_rand_samples=1 and sim_param_set of size 1. Using consistent code paths is generally preferable.

File: naive_joint_space_planner.py [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+        self.eps = eps
+        self.dx, self.dy, self.dz = dx, dy, dz
+        # discretize continuous state space
+        self.dpos = np.array([dx, dy, dz])
+
+        # define action space
+        self.da = da_rad
+        self.num_joints = env.arm.num_joints
+        self.A = ActionSpace(num_DOF=self.num_joints, da_rad=self.da)
+        self.G = dict()
```
Comment (SamNPowers) [inline naive_joint_space_planner.py:139]: I guess I'll stop commenting on this, but short variable names are never my preference, particularly 1 or 2 letter ones. What is G? I can only guess.

File: naive_joint_space_planner.py [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+        if sim_mode == self.SINGLE:
+            self.sim_func = self.env.run_sim
+            assert(self.num_rand_samples == 1)
+        elif sim_mode == self.AVG:
+            self.sim_func = self.env.run_multiple_sims
+            self.process_multiple_sim_results = self.avg_results
+            # 1 is ok, can change, but to double-check that we're not using default value
+            assert(self.num_rand_samples > 1)
+        else:
+            print("Invalid sim mode specified: {}, defaulting to SINGLE".format(sim_mode))
```
Comment (SamNPowers) [inline naive_joint_space_planner.py]: Generally preferable to just bail out entirely if there's something as unexpected as this, rather than quietly covering for the user. Preferable to fail fast in unexpected conditions than let bad experiments continue.

File: naive_joint_space_planner.py [hunk 18]
```diff
@@ -0,0 +1,469 @@
...
+        arm_positions = self.env.arm.get_joint_link_positions(
+            self.joint_pose_from_state(self.start))
+        _, nn_joint_i, nn_joint_pos = (
+            self.dist_arm_to_bottle(self.start, arm_positions))
+        open_set = [
+            Node(0, self.start,
+                 nearest_arm_pos_i=nn_joint_i,
+                 nearest_arm_pos=nn_joint_pos)]
+        closed_set = set()
+        self.G = dict()
```
Comment (SamNPowers) [inline naive_joint_space_planner.py]: Not really sure why this is a class variable? Perhaps I'm missing it, but it looks like it's really only used in plan(), and if it is used by a consumer of the planner, maybe it would be better to return it from plan rather than store it on the object? More expected pattern, anyway.

File: naive_joint_space_planner.py [hunk 19]
```diff
@@ -0,0 +1,469 @@
...
+
+                # completely ignore actions that knock over bottle with high
+                # probability
+                if invalid:
+                    continue
+
+                new_arm_positions = self.env.arm.get_joint_link_positions(
+                    next_joint_pose)
+                trans_cost = (self.calc_trans_cost(n, new_arm_positions) +
+                              1 * fall_prob)
```
Comment (SamNPowers) [inline naive_joint_space_planner.py]: 1 * ? (I assume the 1 was at some point something non-1, but regardless it seems like it should be named and defined, possibly in the constructor with your other planner consts?

File: naive_joint_space_planner.py [hunk 20]
```diff
@@ -0,0 +1,469 @@
...
+                # build next state and check if already expanded
+                next_state = np.concatenate([next_bottle_pos, next_joint_pose])
+                next_state_key = self.state_to_key(next_state)
+                if next_state_key in closed_set:  # if already expanded, skip
+                    continue
+
+                arm_bottle_dist, nn_joint_i, nn_joint_pos = (
+                    self.dist_arm_to_bottle(next_state, new_arm_positions))
+
+                # Quick FIX: use EE for transition costs so set nn_joint to EE
```
Comment (SamNPowers) [inline naive_joint_space_planner.py]: I've seen "EE" a lot but I do not know what it means. End effector? Preferable to be clear about it, for others reading.

**Review by SamNPowers — COMMENTED** [review]
Still not done, maybe about half. Should finish the rest tonight or maybe tomorrow.

File: environment.py [hunk 21]
```diff
@@ -0,0 +1,422 @@
...
+            self.bottle.bottle_id)
+
+        # remove bottle object, can't just reset pos since need to change params each iter
+        p.removeBody(self.bottle.bottle_id)
+
+        # , executed_traj
+        return is_fallen, is_collision, self.bottle.pos, self.bottle.ori, self.arm.joint_pose
+
+    def simulate_plan_online(self, init_joints, policy, bottle_pos, bottle_ori, sim_params: EnvParams):
+        """Run simulation with given joint-space trajectory. Does not reset arm
```
Comment (SamNPowers) [inline environment.py]: simulate_plan and simulate_plan_online seem to have the same doc string, though they do not appear to be doing quite the same thing - mostly seems like a difference of policy specification? (delta joints that get linspace-smoothed, vs just straight commanding the targets?) If I'm right about the difference, not sure "online" is the word I would use to convey that difference? (also the args/returns are out of date)

File: environment.py [hunk 22]
```diff
@@ -0,0 +1,422 @@
...
+        avg_quat = VT[0, :]
+        return avg_quat  # / np.linalg.norm(avg_quat)
+
+    @ staticmethod
+    def state_to_str(state):
+        s = ", ".join(["%.3f" % val for val in state])
+        return s
+
+
+def test_environment_avg_quat():
```
Comment (SamNPowers) [inline environment.py:385]: This is begging to be put into a unit test.

File: main.py (hunk 10 above)
Comment (SamNPowers) [reply main.py]: Looking a bit more, not sure why there's so much environment construction stuff not in the environment definition itself?

File: environment.py [hunk 23]
```diff
@@ -0,0 +1,422 @@
...
+        self.bottle_fric = bottle_fric
+        self.bottle_fill_prob = bottle_fill_prob
+        self.bottle_fric_prob = bottle_fric_prob
+
+    def __repr__(self):
+        return "fill, fric, pfill, pfric: %.2f, %.2f, %.2f, %.2f" % (
+            self.bottle_fill, self.bottle_fric, self.bottle_fill_prob, self.bottle_fric_prob)
+
+
+class Environment(object):
```
Comment (SamNPowers) [inline environment.py:96]: Consider if you were going to release this environment so that others could use it (to write their own methods to compare to yours, for instance), what would the easiest API be for them to use? I personally quite like the OpenAI gym spec, but whatever, as long as it's usable. (Where my definition of "environment" here also generally includes reward definition, i.e. target specification.) In your experiment_helpers you have a function that looks like it's generating a goal? AFAICT that should maybe be in Environment. And you have some object definitions in main.py - AFAICT that should also be in environment. And maybe have Environment only have a step function (and reset etc), and a different SimulationRunner or something class that knows how to pass trajectories of actions into the Environment. 

So basically Environment is responsible for its creation (objects and target) and whatever the "core" functionalities are that the env needs. And then for stuff that you expect others to be less interested in using (your trajectory running code, perhaps) you have a separate object. That also has the advantage of you can have multiple different types of trajectory runners, since it looks like you have several subtypes going on within this code (run_sim vs run_multiple_sims, or "online" vs not-online, or whatever makes the most sense).

File: joint_space_mdp.py [hunk 24]
```diff
@@ -0,0 +1,482 @@
...
+        # color_vals = np.random.randint(0, 0xFFFFFF, size=n)  # +1 for target
+        # colors = [('#%06X' % v) for v in color_vals]
+        for (_, start, end) in sim_log:
+            print(start, end)
+            plt.plot([start[0], end[0]], [start[1], end[1]])
+
+        plt.legend([str(action) for action in self.A], loc='upper left')
+        plt.show()
+
+    def test_reach_p(self):
```
Comment (SamNPowers) [inline joint_space_mdp.py:237]: More good places for unit tests! (Btw if you want advice on how best to do unit tests in python, let me know.)

File: joint_space_mdp.py [hunk 25]
```diff
@@ -0,0 +1,482 @@
...
+            start=self.ymin,
+            stop=self.ymax+self.dy,
+            step=self.dy)
+        self.H, self.W = len(self.Y), len(self.X)
+
+        self.OUT_OF_BOUNDS_COST = MDP.INF
+        self.V = dict()
+        self.P = dict()
+
+    def state_to_key(self, state):
```
Comment (SamNPowers) [inline joint_space_mdp.py:49]: "key"? Quite general

File: joint_space_mdp.py [hunk 26]
```diff
@@ -0,0 +1,482 @@
...
+            for ai in best_actions:
+                action = self.A[ai]
+                # if desired x,y,z is out of reach, skip this action
+                # validity depends on contact height
+                target_dist = np.linalg.norm(np.array([x, y]))
+                if target_dist > self.env.arm.calc_max_horiz_dist(
+                        action.height):
+                    continue
+
+                value, ns = self.env.run_sim_stochastic(action)
```
Comment (SamNPowers) [inline joint_space_mdp.py:106]: Where is run_sim_stochastic defined? It looks like it's expected on Environment, but doesn't seem to be there?

File: joint_space_mdp.py [hunk 27]
```diff
@@ -0,0 +1,482 @@
...
+            self.P[yi][xi] = best_actions
+            end = time.time()
+            print("Time(s) for one state: %.3f" % (end - start))
+            total_time += (end-start)
+        print("Total Runtime of Update Policy: %.3f" % total_time)
+
+    def init_action_space(self, run_full_mdp):
+        A = []  # need to maintain order
+        self.da = math.pi/80
+        if run_full_mdp:
```
Comment (SamNPowers) [inline joint_space_mdp.py:194]: Not sure why this boolean would map to what's happening in the body of the if/then. Comment? (Btw I normally think of variable names and code logic as "what" and comments as "why" - so if the why is clear, no comments necessary, but in this case I find the why not very clear.)

**Review by SamNPowers — COMMENTED** [review]
I'm so sorry this took me so long to look through all of (rebuttal involved more work than I thought). 

Overall I found it quite readable, and generally, I think, understandable.

Here's what my biggest piece of feedback is: consider separating things out as if you plan on re-using them (or plan on releasing it for others to use). 

If someone was going to develop their own method using this bottle-arm environment, how easy would that be for them? (Would they have to do a whole bunch of target/bottle/arm initialization on their own?)

If you want to apply your planner to something that isn't this bottle-arm environment, how easy would *that* be? "Bottle" and "arm" specific stuff appear all over the place. Perhaps you create a BottleArmEnvironment, then maybe a Planner class with a BottleArmPlanner subclass, where generic stuff goes in Planner, and then specific stuff goes in the subclass (e.g. compute_heuristic) or whatever. Then when you want to apply your planning logic to some other environment all the shared stuff would be re-usable, and you'd only have to specify the bare minimum necessary. It seems like NaivePlanner is aaalmost there in terms of this, but then a bunch of arm-bottle specific stuff occurs in lots of places, e.g. in main.py. Consolidate it so it's easier to change out, and your future self will thank you.

Unit tests help with having clear APIs. Define what you want your shared, re-usable interface to be, then write a unit test where everything except what you're testing is spoofed. E.g. write a fake Environment that your NaivePlanner can use. That'll quickly give you a good sense for how hard/easy it'll be to re-use your code on a different task. And vice versa, create a simple "policy" that uses the environment (random, perhaps) - how easy is it to set up the Environment?

Let me know if you have any questions, want to hop on a call, etc.

File: sim_objects.py (hunk 1 above)
Comment (Alvinosaur) [reply sim_objects.py]: Makes sense

File: sim_objects.py (hunk 2 above)
Comment (Alvinosaur) [reply sim_objects.py]: Oh wow yes!

File: sim_objects.py (hunk 3 above)
Comment (Alvinosaur) [reply sim_objects.py:29]: Ah yes, I am aware of this convention, but was just lazy about it. I will make this change later since it affects many more things, but it definitely would help me avoid making hacky fixes that directly overwrite the private variables, a sign that I should write a public function to do this.

File: sim_objects.py (hunk 5 above)
Comment (Alvinosaur) [reply sim_objects.py]: That is a great idea, having the get_general_max_reach would also be good, and both could make use of the example syntax:

temperature = property(get_temperature, set_temperature)

File: main.py (hunk 6 above)
Comment (Alvinosaur) [reply main.py]: Oh yes, I will move all these constants to the top of the script so I know where to look next time. A lot of the bugs I saw in the past did involve adding hacky changes to these constants halfway in the code. I would see the initial definition of a constant, but the constant would unexpectedly change later on. By forcing everything to be defined at the top, this bug won't happen.

File: main.py (hunk 7 above)
Comment (Alvinosaur) [reply main.py:1]: Oh definitely, for this particular project I didn't expect anyone else to use it which is why I didn't do extensive testing, but I honestly should have just done this since it would fix many bugs. These comments are great and I should just follow these best practices regardless of whether people will use the code. 

Yea a requirements file would be good.

File: main.py (hunk 8 above)
Comment (Alvinosaur) [reply main.py]: Yes you are completely correct. I wanted the environment to have be responsible for generating environment physics parameters, but I also wanted to give the planner flexibility in choosing which physics parameters to use during planning in case I wanted to do some experiments. In hindsight, I think I should just make the environment solely responsible though and just call some environment function to do this.

File: main.py (hunk 9 above)
Comment (Alvinosaur) [reply main.py]: Agreed
